### PR TITLE
[IMP] survey: improve sessions usability and design

### PR DESCRIPTION
**Pull request by awa-odoo**
PURPOSE

Improve the various sessions flow to make it easier for attendees to reach the
session and to allow better interactions between the host and his audience.

This comes with small usability improvements as well as increased
"beautifulness" by using custom layouts & animations.

SPECS

General specifications, see sub-commits for details.

- Currently, the "short link" feature is good but it still needs to be improved
  with the following points:
    - Make the default "code" shorter (4) and digit only.
    - Allow to customize the "code"
    - Check for potential collisions
- When attendees reach the survey through the code, they need to land on the
  "waiting screen" even if the session is not started yet. This will allow the
  host to show the survey link even if he forgets to click on "start session"
  first.
- Reduce the fade out / fade in delay when moving from one question to another.
  It's currently 2 seconds, change it to 1 second.
  Right now the delay gives a "slowness" impression.
- Display matrix choices on the host screen.
  It will allow attendees that are not participating to have more context.
- Add python test/tours to test the survey session flow from the host point of view
- Improve survey form navigation (moved buttons, added keyboard navigation)
- Auto submit questions and matrix with simple choice

DESIGN TWEAKS

Slightly improve the survey form by adding a few cosmetic changes,
notably:

- Add/reduce some margins/paddings
- Adjust colors/font-weight
- Adjust heights
- Change some icons by more fitting replacements
- Center matrix choices
- Move survey progression to bottom-right
- Remove ugly jumbotrons
- Hide survey title in page_per_question layout
- When there are 5 or less option for simple_choice or multiple_choice
  questions, display the options vertically.
- Only show the textarea for the "comment answer" if the user clicks on the
  related option.

LINKS

Upgrade PR odoo/upgrade#897
Task 2208574

### CODE REVIEW

File: addons/survey/controllers/main.py [hunk 1]
```diff
@@ -311,10 +311,29 @@ def _prepare_survey_data(self, survey_sudo, answer_sudo, **post):
...
+                })
+            elif survey_sudo.questions_layout == 'page_per_question':
+                page_ids = survey_sudo.question_ids.ids
+                result['survey_progress'] = request.env.ref('survey.survey_progression').render({
+                    'survey': survey_sudo,
+                    'page_ids': page_ids,
+                    'page_number': page_ids.index(survey_data['question'].id)
+                })
+
+        return result
```
Comment (tde-banana-odoo) [inline addons/survey/controllers/main.py]: [C5] Would maybe be easier to read by computing survey_content, survey_progress, and return {'survey_content': survey_content, 'survey_progress': survey_progress}

File: addons/survey/controllers/main.py [hunk 2]
```diff
@@ -311,10 +311,29 @@ def _prepare_survey_data(self, survey_sudo, answer_sudo, **post):
...
+            result = {'survey_content': request.env.ref('survey.survey_fill_form_done').render(survey_data)}
+        else:
+            result = {'survey_content': request.env.ref('survey.survey_fill_form_in_progress').render(survey_data)}
+
+        if answer_sudo.state == 'in_progress' and not survey_data.get('question', request.env['survey.question']).is_page:
+            if survey_sudo.questions_layout == 'page_per_section':
+                page_ids = survey_sudo.page_ids.ids
+                result['survey_progress'] = request.env.ref('survey.survey_progression').render({
+                    'survey': survey_sudo,
+                    'page_ids': page_ids,
```
Comment (tde-banana-odoo) [inline addons/survey/controllers/main.py]: [C5] We could directly prepare the dict, and include the condition like
```
'page_ids': survey_sudo.page_ids.ids if survey_sudo.questions_layout == 'page_per_section' else survey_sudo.question_ids.ids
'page_number': blah
```

That way we have less embedded if / else .

File: addons/survey/static/src/js/survey_form.js [hunk 3]
```diff
@@ -55,6 +55,10 @@ publicWidget.registry.SurveyFormWidget = publicWidget.Widget.extend({
                 self.preventEnterSubmit = true;
             }
             self._initSessionManagement();
+
+            // Needs global selector as the progress is not within the survey form, but needs to be
+            // updated at the same time
+            self.$surveyProgress = $('.o_survey_progress_wrapper');
```
Comment (tde-banana-odoo) [inline addons/survey/static/src/js/survey_form.js]: [C5] A bit strange to update html outside of the widget's scope. I suppose having another level (survey_page with a form and a progress) would require too much refactoring ?

File: addons/survey/controllers/main.py [hunk 4]
```diff
@@ -315,7 +315,10 @@ def _prepare_question_html(self, survey_sudo, answer_sudo, **post):
         if answer_sudo.state == 'done':
             result = {'survey_content': request.env.ref('survey.survey_fill_form_done').render(survey_data)}
         else:
-            result = {'survey_content': request.env.ref('survey.survey_fill_form_in_progress').render(survey_data)}
+            result = {
+                'survey_content': request.env.ref('survey.survey_fill_form_in_progress').render(survey_data)
+            }
+        result['survey_navigation'] = request.env.ref('survey.survey_navigation').render(survey_data)
```
Comment (tde-banana-odoo) [inline addons/survey/controllers/main.py]: [C6] Same remark here, include it in final result dict.

File: addons/survey/controllers/main.py [hunk 5]
```diff
@@ -315,7 +315,10 @@ def _prepare_question_html(self, survey_sudo, answer_sudo, **post):
         if answer_sudo.state == 'done':
             result = {'survey_content': request.env.ref('survey.survey_fill_form_done').render(survey_data)}
         else:
-            result = {'survey_content': request.env.ref('survey.survey_fill_form_in_progress').render(survey_data)}
+            result = {
+                'survey_content': request.env.ref('survey.survey_fill_form_in_progress').render(survey_data)
```
Comment (tde-banana-odoo) [inline addons/survey/controllers/main.py]: [C6] unnecessary diff

File: addons/survey/views/survey_templates.xml [hunk 6]
```diff
@@ -597,5 +588,31 @@
             </div>
         </t>
     </template>
+
+    <template id="survey_navigation" name="Survey: Navigation">
```
Comment (tde-banana-odoo) [inline addons/survey/views/survey_templates.xml]: [C6] linting: I would move it to survey_templates_management.xml (in tools / utilities), to keep main templates only here (already a big file)

File: addons/survey/views/survey_templates.xml [hunk 7]
```diff
@@ -597,5 +588,31 @@
...
+        -->
+        <t t-set="page_per_section_back_button" t-value="can_go_back and survey.questions_layout == 'page_per_section' and page != survey.page_ids[0]" />
+        <!-- Can go back in 'page_per_question' if:
+            - Back button is allowed in configuration
+            - This is not a session answer (doesn't make sense to go back in session context)
+            - This is not the first question
+            - The survey doesn't have pages OR this is not the first page of the survey
+                (pages are displayed in 'page_per_question' layout when they have a description, see PR#44271)
+        -->
+        <t t-set="page_per_question_back_button"
```
Comment (tde-banana-odoo) [inline addons/survey/views/survey_templates.xml]: [C6] At one point it becomes easier to prepare values in controllers than trying to understand complex variable value in template. I would move it to the render, especially that it is already rendered server-side in a specific render call.

File: addons/survey/views/survey_templates.xml [hunk 8]
```diff
@@ -21,12 +21,25 @@
             <div id="wrap" class="oe_structure oe_empty"/>
         </xpath>
         <xpath expr="//footer" position="after">
-            <div class="py-3">
-                <div class="float-right px-5 py-2 bg-white rounded-left">
-                    <t t-call="web.brand_promotion_message">
+            <div class="py-3 m-0 p-0 text-right">
+                <div class="o_survey_progress_wrapper d-inline-block pr-5 text-left">
+                    <t t-call="survey.survey_progression"
```
Comment (tde-banana-odoo) [inline addons/survey/views/survey_templates.xml]: Why are you calling survey_progression that is also rendered by a route and added by JS ? Is it for some initialization ?

File: addons/survey/views/survey_templates.xml [hunk 9]
```diff
@@ -166,42 +182,44 @@
...
-                        </t>
-                    </div>
+                        Continue</button><span class="font-weight-bold text-muted ml-2 d-none d-md-inline"> or press Enter</span>
                 </div>
             </div>
         </t>
 
-        <t t-if="survey.questions_layout == 'page_per_question'">
+        <!-- If we have a choice question and less than 6 options, we want minimized display.
+        Minized display means we display the choices vertically (instead of optimized based on screen space).
```
Comment (tde-banana-odoo) [inline addons/survey/views/survey_templates.xml]: [C5] Minized -> Minimized

File: addons/survey/models/survey_survey.py [hunk 10]
```diff
@@ -563,6 +563,29 @@ def _get_next_page_or_question(self, user_input, page_or_question_id, go_back=Fa
...
+        If yes, the given page is not the last one.
+        """
+        pages_or_questions = self._get_pages_or_questions(user_input)
+        current_page_index = pages_or_questions.ids.index(page_or_question.id)
+        next_page_or_question_candidates = pages_or_questions[current_page_index + 1:]
+        is_last_question = True
+        if next_page_or_question_candidates:
+            inactive_questions = user_input._get_inactive_conditional_questions()
+            if self.questions_layout == 'page_per_question':
+                is_last_question = not any(next_question not in inactive_questions for next_question in next_page_or_question_candidates)
```
Comment (tde-banana-odoo) [inline addons/survey/models/survey_survey.py]: [C7] Why keeping a is_last_question variable ? You could make direct returns like
```
if not next_page_or_question_candidates: return False
...
return not any(blah) 
...
```

Just sayin'

File: addons/survey/models/survey_survey.py [hunk 11]
```diff
@@ -23,6 +23,30 @@ class Survey(models.Model):
...
+        """ Attempt to generate a session code for our survey.
+        The method will first try to generate 10 codes with 4 digits each and check if any are colliding.
+        If we have at least one non-colliding code, we use it.
+        If all 10 generated codes are colliding, we try with 10 codes of 5 digits,
+        then 6, ... up to 10 digits. """
+
+        for digits_count in range(4, 10):
+            range_lower_bound = 1 * (10 ** (digits_count - 1))
+            range_upper_bound = (range_lower_bound * 10) - 1
+            code_candidates = set([str(random.randint(range_lower_bound, range_upper_bound)) for i in range(10)])
```
[the remark this answers is not in the document]
Comment (awa-odoo) [reply addons/survey/models/survey_survey.py]: Ok for 20 (19 was not enough, and 21 too much)

File: addons/survey/models/survey_survey.py [hunk 12]
```diff
@@ -6,6 +6,8 @@
 import uuid
 import werkzeug
 
+from dateutil.relativedelta import relativedelta
```
[the remark this answers is not in the document]
Comment (awa-odoo) [reply addons/survey/models/survey_survey.py]: Lost during rebase hell probably, I removed it.

File: addons/survey/controllers/main.py (hunk 2 above)
Comment (awa-odoo) [reply addons/survey/controllers/main.py]: I like the idea, but I tried and it's actually a lot less readable.
Mainly because of the fact that we re-use page_ids for the page_number.

File: addons/survey/static/src/js/survey_form.js (hunk 3 above)
Comment (awa-odoo) [reply addons/survey/static/src/js/survey_form.js]: The problem is that the progress / navigation are now included in the generic "survey.layout" template.

So the widget would have to be plugged really high to include all that (probably something like "#wrap") which is most likely not a good idea.

I agree that this needs to be refactored.
We need to sit down and think about how we could organize those templates better in terms of structure.

File: addons/survey/views/survey_templates.xml (hunk 8 above)
Comment (awa-odoo) [reply addons/survey/views/survey_templates.xml]: It is rendered by the JS side only when you submit, because the page is not fully refreshed (AJAX magic + fade out / fade in of next question -> we need to update the progression).

But it also needs to work when you fully refresh the page, then the template is rendered and the JS does not handle anything (exact same mechanic as the form content).

File: addons/survey/controllers/main.py [hunk 13]
```diff
@@ -268,10 +268,12 @@ def _prepare_survey_data(self, survey_sudo, answer_sudo, **post):
...
             previous_page_or_question_id = int(post['previous_page_id'])
             new_previous_id = survey_sudo._get_next_page_or_question(answer_sudo, previous_page_or_question_id, go_back=True).id
+            page_or_question = request.env['survey.question'].sudo().browse(previous_page_or_question_id)
             data.update({
-                page_or_question_key: request.env['survey.question'].sudo().browse(previous_page_or_question_id),
+                page_or_question_key: page_or_question,
                 'previous_page_id': new_previous_id,
-                'has_answered': answer_sudo.user_input_line_ids.filtered(lambda line: line.question_id.id == new_previous_id)
+                'has_answered': answer_sudo.user_input_line_ids.filtered(lambda line: line.question_id.id == new_previous_id),
+                'can_go_back': survey_sudo._can_go_back(answer_sudo, page_or_question)
```
Comment (tde-banana-odoo) [inline addons/survey/controllers/main.py]: Linting: add a final ",", so that if in future changes we add something in the dict we don't update that line

File: addons/survey/controllers/main.py [hunk 14]
```diff
@@ -296,7 +298,8 @@ def _prepare_survey_data(self, survey_sudo, answer_sudo, **post):
 
             data.update({
                 page_or_question_key: next_page_or_question,
-                'has_answered': answer_sudo.user_input_line_ids.filtered(lambda line: line.question_id == next_page_or_question)
+                'has_answered': answer_sudo.user_input_line_ids.filtered(lambda line: line.question_id == next_page_or_question),
+                'can_go_back': survey_sudo._can_go_back(answer_sudo, next_page_or_question)
```
Comment (tde-banana-odoo) [inline addons/survey/controllers/main.py]: Linting: same here

File: addons/survey/controllers/main.py [hunk 15]
```diff
@@ -338,7 +341,8 @@ def _prepare_question_html(self, survey_sudo, answer_sudo, **post):
 
         return {
             'survey_content': survey_content,
-            'survey_progress': survey_progress
+            'survey_progress': survey_progress,
+            'survey_navigation': request.env.ref('survey.survey_navigation').render(survey_data)
```
Comment (tde-banana-odoo) [inline addons/survey/controllers/main.py]: Linting: same here :D

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-merge
